_posts/2016-08-31-continuous-delivery-databases.md

```diff
@@ -0,0 +1,57 @@
+Nacharbeiten aufgrund schlechter Qualität.  
+
+## Lösungsmöglichkeit
+
+Wir verwenden eine komplett auf Docker basierende Continuous Delivery Pipeline. Dies bedeutet, dass wir ab einem bestimmten Schritt nicht
+mehr mit den jar / war Artefakten arbeiten, sondern dirket mit den erzeugten Docker Images.
+
+Unter diesen Randbedingungen sieht unsere Lösung in etwa wie folgt aus:
+
+* Wir verwenden Liquibase, da das Tool datenbankunabhängig ist und wir damit einfach MySQL, unterschiedliche Kontexte und Basisdaten verwalten können.
+* Jede Applikation / Service enthält ein Modul "dbmigrations", welches alle notwendigen Liquibase Migrationen enthält (als SQL Dateien).
+* In einem bestimmten Schritt erzeugen wir die entsprechenden Docker Images, unter anderem dabei auch das Liquibase Image. Dazu holen wir die dbmigrations Module aus dem Maven Repository. Dieses Image beinhaltet ein Skript, welches die DB Migrations dann auch auf eine Zieldatenbank anwenden kann.
+* Es gibt einen weiteren Pipelineschritt, der frühzeitig feststellen soll, ob die Migrations "fehlerfrei" sind und auf eine bestehende Datenbank angewendet werden können. Dazu holen wir das Liquibase DB Migrations Image, welches aktuell auf der UAT Umgebung deployed wurde. Dieses wenden wir auf eine "leere" Datenbank an. Anschließend wird das aktuell gebaute Liqubiase Image auf dieser Datenbank angewendet. Dies setzt voraus, dass niemand manuell Änderungen an der Datenbank durchführt. Ansonsten sollte man immer das DDL direkt aus der Zieldatenbank (UAT, Prod) erzeugen, gegen welche man testen will.
+* Das Deployment erfolgt in zwei Schritten. Die Umgebungen sind als docker-compose Dateien beschrieben. Docker Compose reicht aber alleine nicht aus, um ein erfolgreiches Deployment sicher zu stellen. Dies liegt daran, dass Docker Compose abhängige Container direkt startet, wenn die Container, die als Abhängigkeit definiert sind, gestartet sind. Gestartet heißt aber nicht, dass der darin enthaltene Prozess wirklich "fertig" ist. Wir lösen dies, indem wir das docker-compose File mittels eines Shell Skripts nur bis zum liquibase Container starten und warten, dass sich dieser wieder (erfolgreich) beendet. Dann wissen wir, dass die Datenbank im neuen Zustand vorhanden ist und die Applikation starten kann.
+
+### Diskussion
+Einige Leser werden sich sicher fragen, warum wir für die Schemamigration einen eigenen Container verwenden und dies nicht direkt von der Anwendung ausführen lassen.
+Der Grund liegt darin begründet, dass wir die Migration unabhängig von der Anwendung ausführen wollen.
+Dies ist wichtig, um z.B. Skripte zur Datenbankbefüllung während der Entwicklungs- und Test-Phasen zu testen, welche die Datenbank befüllen. Diese Tests laufen teilweise auf Datenbanken, auf denen keine Anwendung zugreift.
```

Comment:
Der Nebensatz `, welche die Datenbank befüllen.` ist redundant zu `Skripte zur Datenbankbefüllung`.

Im zweiten Satz `denen` durch `die` ersetzen. Klingt in meinen Augen holprig.